### ISSUE TITLE
pkt-gen or netmap not working on veth interfaces

### ISSUE THREAD

**Issue by Hesam-Naderi-Dehkordi** [luigirizzo/netmap#118]
I have spend weeks of time on search of how to do a high speed packet generation in the internet. I found moongen,pf_ring,... but found that your framework maybe my only chose, because my setup is in virtual network and i need high speed packet generation on veth interfaces.
I have successfully build netmap and driver and modules, and loaded them. for example output for veth module built with the framework and copied to the kernel drivers directory :
modinfo veth
filename:       /lib/modules/3.16.7-29-desktop/kernel/drivers/net/veth.ko
alias:          rtnl-link-veth
license:        GPL v2
description:    Virtual Ethernet Tunnel
srcversion:     1AE927BF04FA0D8C8752F49
depends:        netmap
vermagic:       3.16.7-29-desktop SMP preempt mod_unload modversions 

I mean that I think made netmap in the correct way but when I try some test with pkt-gen I get the following results :

 ./pkt-gen -i eth0 -f tx -n 500111222 -l 60 -w 5
039.779148 main [1930] interface is eth0
039.779283 main [2050] running on 1 cpus (have 8)
039.779946 extract_ip_range [367] range is 10.0.0.1:0 to 10.0.0.1:0
039.779962 extract_ip_range [367] range is 10.1.0.1:0 to 10.1.0.1:0
039.780244 main [2148] mapped 334980KB at 0x7f622fb9f000
Sending on netmap:eth0: 1 queues, 1 threads and 1 cpus.
10.0.0.1 -> 10.1.0.1 (00:00:00:00:00:00 -> ff:ff:ff:ff:ff:ff)
039.780277 main [2233] Sending 512 packets every  0.000000000 s
039.780283 main [2235] Wait 5 secs for phy reset
044.780965 main [2237] Ready...
044.781304 sender_body [1211] start, fd 17 main_fd 17
045.783182 main_thread [1720] 1.021 Kpps (1.023 Kpkts 491.040 Kbps in 1001978 usec) 511.50 avg_batch 0 min_space
046.784829 sender_body [1280] poll error/timeout on queue 0: Success
046.784854 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001675 usec) 0.00 avg_batch 99999 min_space
047.786888 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1002034 usec) 0.00 avg_batch 99999 min_space
048.786863 sender_body [1280] poll error/timeout on queue 0: Success
048.788286 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001398 usec) 0.00 avg_batch 99999 min_space
049.789788 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001502 usec) 0.00 avg_batch 99999 min_space
^C050.517705 sigint_h [404] received control-C on thread 0x7f6244c7c700
050.517759 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 727971 usec) 0.00 avg_batch 99999 min_space
050.789891 sender_body [1326] flush tail 1023 head 1023 on thread 0x7f622fb9e700
050.789932 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
050.790039 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
050.790884 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
050.791880 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
050.792890 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.000157 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.001141 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.002160 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.003147 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.004161 sender_body [1334] pending tx tail 1023 head 1023 on ring 0
051.519753 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001994 usec) 0.00 avg_batch 99999 min_space

no packet is send at all. Can you guide me what's wrong?Or maybe it's a bug?

my setup is linux 3.16.7 (opensuse 13.2), netmap downloaded from master branch and virtual switch and hosts are made by mininet and orginal openvswitch (newest,not patched to the old days 2012).

**Comment by vmaffione** [comment]
Hi,
  It is not clear whether you are using the netmap-patched driver for veth or not.
Have you loaded the veth.ko module compiled in the LINUX directory?

If you are not using the patched driver for veth, you should try to load netmap in the following way

# insmod netmap.ko generic_txqdisc=1

When not using the patched veth driver you should see packet rates in the order of 1-3 Mpps.
When using the patched veth driver you should see packet rates in the order of 80-120 Mpps.

**Comment by Hesam-Naderi-Dehkordi** [comment]
Hi again.
Yes I think that it's shown in the lsmod that it depends on netmap and so
it's the patched one.
Indeed I have overwritten the orginal module in kernel module paths with
the patched modules build with netmap.
But I see no packets are generated. I can confirm this by checking
statistics with ifconfig command so that no increase after execution
of  pkt-gen
-i eth0 -f tx -n 500111222 -l 60 -w 5
I have also tried your direction insmod netmap.ko generic_txqdisc=1 but
nothing changes.
getting just these messages only :
044.780965 main [2237] Ready...
044.781304 sender_body [1211] start, fd 17 main_fd 17
045.783182 main_thread [1720] 1.021 Kpps (1.023 Kpkts 491.040 Kbps in
1001978 usec) 511.50 avg_batch 0 min_space
046.784829 sender_body [1280] poll error/timeout on queue 0: Success
046.784854 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001675
usec) 0.00 avg_batch 99999 min_space
047.786888 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1002034
usec) 0.00 avg_batch 99999 min_space
048.786863 sender_body [1280] poll error/timeout on queue 0: Success
048.788286 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001398
usec) 0.00 avg_batch 99999 min_space
049.789788 main_thread [1720] 0.000 pps (0.000 pkts 0.000 bps in 1001502
usec) 0.00 avg_batch 99999 min_space
....


On Wed, Dec 16, 2015 at 5:05 PM, Vincenzo Maffione <notifications@github.com

**Comment by vmaffione** [comment]
It may be that veth patched driver is somewhat broken in your kernel version or in your environment.

However, the txqdisc parameter has no effect because you keep using the patched one. At this point, try to load the netmap module also passing the parameter admode=2. In this way you force netmap to ignore that the driver is patched.

Alternatively, you can change the netmap parameters through /sys/module/netmap/parameters, without the need to reload the module.

If you mean to run the transmitter indefinitely, you can just drop the -n argument. Moreover, the -w argument should only be useful with hardware device, not with software ones (like veth), so you can omit it.

**Comment by Hesam-Naderi-Dehkordi** [comment]
I have captured and record my screen, here is the upload link:
http://filepi.com/i/TrMhksa
Can you please watch it and help me ? I really spend weeks to found netmap and now it's not working :(

Another bug is that if there be a dash in the interface name netmap wouldn't bound to it , because of this I changed the interface name in the recording.

**Comment by vmaffione** [comment]
1) Make sure you have the most updated version of netmap.

2) Reboot the machine and start with this
````
# modprobe netmap generic_txqdisc=1 admode=2
# modprobe veth
````

**Comment by Hesam-Naderi-Dehkordi** [comment]
Okay thanks. I grabed netmap 2 or 3 days ago, but I 'll  grab another copy
from master branch of github now and build it and try to make screen
capture again. I 'll inform you in 1-2 hours.

Thank you so much :)

On Wed, Dec 16, 2015 at 7:08 PM, Vincenzo Maffione <notifications@github.com

**Comment by Hesam-Naderi-Dehkordi** [comment]
for number 1)I have cloned master branch again, builded it , but no change. here is the uploaded screen capture:
http://filepi.com/i/yAnuvPJ

and for number 2) I restarted the system, did what you said, Indeed put mininet away and worked with pure veth created by ip link but no changes again. Here is the screen capture link :
http://filepi.com/i/vB4SCKD

**Comment by vmaffione** [comment]
2015-12-16 18:01 GMT+01:00 Hesam-Naderi-Dehkordi <notifications@github.com>:

> for number 1)I have cloned master branch again, builded it , but no
> change. here is the uploaded screen capture:
> http://filepi.com/i/yAnuvPJ
>

Steps seem ok.
Now I recall that with the veth patched driver you need both veth0 and
veth1 to be up **and** in netmap mode.
Try to start first the receiver "pkt-gen -f rx -i veth0" and then the
transmitter "pkt-gen -f tx -i veth1".

> and for number 2) I restarted the system, did what you said, Indeed put
> mininet away and worked with pure veth created by ip link but no changes
> again. Here is the screen capture link :
> http://filepi.com/i/vB4SCKD
>

Check the spelling of the parameters, they are wrong.
Have you ifconfig up both veth0 and veth1? It's not required to assign an
IP address, but you can do it, ofc.
Check in /sys/module/netmap/parameters that the parameters are ok.

Try also here to start first the receiver, then the transmitter.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luigirizzo/netmap/issues/118#issuecomment-165175293>.
>



-- 
Vincenzo Maffione

**Comment by vmaffione** [comment]
In any case, take into account that veth patched driver is experimental, and possibly unsafe. We're working on that.

**Comment by Hesam-Naderi-Dehkordi** [comment]
Hello again.
pkt-gen -i veth0 -f tx
pkt-gen -i veth1 -f rx

I tried what you said and it works. The other pair of veth needs to be up
and running pkt-gen -i -f rx. It means that the other pair must be in
receiving mode so that the veth pair start sending.

when running in patched mode  it gives about 5Mpps and when running in
original mode gives about  250kps. these are tests in pure veth setup
(without any switches in between)

But there is still some problem.
The problem is that in my setup veth end is connected as a port of
openvswitch. In other words there are two hosts connected together with an
openvswitch and links are veth pairs.

I put the other end of veth 1 in receiving mode ( by using pkt-gen -i
veth1-1 -f rx) and try to produce some traffic by veth1-0  on host H1 with
src ip 10.0.0.1 and dstip 10.0.0.2 but no traffic is received at host H2.
It seems that all traffic generated is consumed by end of veth1-2 pair and
never receives to switch. And if I don't run "pkt-gen -i veth1-1 -f rx" no
traffic is able to produce by "pkt-gen -i veth0 -f tx"
I have attached a basic schema of my setup. maybe it helps.
If I don't use netmap mode the traffic is able to reach H2, but it's
throughput and performance is poor and really need netmap for this approach.


On Fri, Dec 18, 2015 at 1:42 AM, Vincenzo Maffione <notifications@github.com

**Comment by Hesam-Naderi-Dehkordi** [comment]
![basic schema](https://cloud.githubusercontent.com/assets/8275634/11899705/26c18cd4-a5b6-11e5-99ed-a2872e0fddf5.jpg)

**Comment by vmaffione** [comment]
In your setup you cannot use the patched veth driver, because openvswitch does not know how to use netmap.
However, you can use the unpatched driver. I replicated this setup and it works to me, even if the performances are not great (~600 Kpps).

To overcome performance problems fo this setup, try to use the VALE switch in place of OVS.

In any case, I recommend to use this commit (791d1c88c173a5ac4413b3e447fe0c1feab224fc), since current master (05b95d8d68fee2245bd22fb73131efa140c8d706) has some temporary problems introduced by the last Pull Request.

**Comment by vmaffione** [comment]
Current master (519c07f2b2fa8774fcfb232d262701ffdc5a1682) it's ok again.

**Comment by joerango** [comment]
I am having similar issues with netmap + veth on Linux.

I am trying to do some tests using ping and iperf on the topology below:
(veth1 : 192.168.150.1 in the root namespace) -------[veth2: vale0 : veth3] ------[veth4: vale1: veth5] ---------(veth6: 192.168.150.2 in a separate network namespace)

vale0 and vale1 are VALE switches created by adding the veth interface using the 'vale-ctl -a' command.

My goal is to use VALE and netmap in network experiments experiencing limited bandwidth so interfaces veth3 and veth4 have a netem qdisc attached to them to limit bandwidth.

Netmap code used is from current master (29319d4). Kernel is 3.18.26.

I have tried 3 different configurations and none of them work properly.

1. netmap module loaded without parameters + unpatched veth driver
  * ping works
  * iperf works for 5-10 seconds then halts. I can force quit it but ping doesn't work anymore.
  
2. netmap module loaded without parameters + patched veth driver
  * ping doesn't work
  
3. netmap module loaded with generic_txqdisc=1 + patched veth driver
  * ping doesn't work
  
4. netmap module loaded with admode=2 + patched veth driver
  * ping works
  * iperf works for 5-10 seconds then halts. I can force quit it but ping doesn't work anymore.
  
4. netmap module loaded with generic_txqdisc=1 and admode=2 + patched veth driver
  * ping works
  * iperf works fine but the netem qdisc is bypassed. Outgoing bandwidth on veth3 and veth4 is not limited.
  
Any ideas how to work around this issue?

**Comment by vmaffione** [comment]
Hi,
  Thanks for reporting.

First of all, I'm going to push some fixes that should address 1, in these
days. So please stay tuned on the master branch.

In reply to your points:

1. Yes, there is a bug that affects veth and bridge interfaces with
unpatched drivers. See above.

2. Unfortunately the veth patched driver is very experimental. So just skip
this option for now, we will investigate on it.

3. generic_txqdisc does not have any effect here, it has effect only when
you use unpatched drivers or you specify admode=2.

4. Using admode=2 you are telling netmap to avoid using the netmap
functionalities in the patched driver and use normal driver functionalities
instead (it does not matter whether the driver is patched or not). So this
configuration is the very same as 1.

5. Yes, this is not a bug. In fact, with generic_txqdisc=1, netmap is
replacing the existing qdisc with a FIFO one that it can control. This is
necessary for correct netmap functionality in general.


The solution should be to wait for my updates and try again (1) or (4) - I
would say (1) it's just easier, because you don't have to insmod a modified
veth module.

Apart from this, I have some additional suggestions.

The point of netmap is bypassing OS network stack and normal driver
datapath.
I'm not sure whether it makes sense in your scenario to use netmap in place
of normal kernel bridges if you use unpatched drivers or admode=2.
In my view, the right way to go with netmap for your use-case would be one
of the following:

======option A=======

(veth1 : 192.168.150.1 in the root namespace) -------[veth2: valeXXX:
veth5] ---------(veth6: 192.168.150.2 in a separate network namespace)

where vale XXX has a custom "bridge forwarding function" that you provide
(different from the default learning switch) that implements rate limiting.

======option B=======
(veth1 : 192.168.150.1 in the root namespace) -------veth2 [bridgeXXX]
veth5 ---------(veth6: 192.168.150.2 in a separate network namespace)

Where bridgeXXX is a modified version of the examples/bridge.c program that
already implements a bridge in userspace, where you should add support for
rate limiting.

option B should be just easier.

2016-02-11 1:26 GMT+01:00 Joseph Beshay <notifications@github.com>:

> I am having similar issues with netmap + veth on Linux.
>
> I am trying to do some tests using ping and iperf on the topology below:
> (veth1 : 192.168.150.1 in the root namespace) -------[veth2: vale0 :
> veth3] ------[veth4: vale1: veth5] ---------(veth6: 192.168.150.2 in a
> separate network namespace)
>
> vale0 and vale1 are VALE switches created by adding the veth interface
> using the 'vale-ctl -a' command.
>
> My goal is to use VALE and netmap in network experiments experiencing
> limited bandwidth so interfaces veth3 and veth4 have a netem qdisc attached
> to them to limit bandwidth.
>
> Netmap code used is from current master (29319d4
> <https://github.com/luigirizzo/netmap/commit/29319d439a56f7b9dc2d72c777cbc826b8a2aafa>).
> Kernel is 3.18.26.
>
> I have tried 3 different configurations and none of them work properly.
>
>    1.
>
>    netmap module loaded without parameters + unpatched veth driver
>    - ping works
>       - iperf works for 5-10 seconds then halts. I can force quit it but
>       ping doesn't work anymore.
>    2.
>
>    netmap module loaded without parameters + patched veth driver
>    - ping doesn't work
>    3.
>
>    netmap module loaded with generic_txqdisc=1 + patched veth driver
>    - ping doesn't work
>    4.
>
>    netmap module loaded with admode=2 + patched veth driver
>    - ping works
>       - iperf works for 5-10 seconds then halts. I can force quit it but
>       ping doesn't work anymore.
>    5.
>
>    netmap module loaded with generic_txqdisc=1 and admode=2 + patched
>    veth driver
>    - ping works
>       - iperf works fine but the netem qdisc is bypassed. Outgoing
>       bandwidth on veth3 and veth4 is not limited.
>
> Any ideas how to work around this issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luigirizzo/netmap/issues/118#issuecomment-182654584>.
>



-- 
Vincenzo Maffione

**Comment by vmaffione** [comment]
You can try with the current master, if you wish. The fix is now upstream.

**Comment by joerango** [comment]
Thanks for your detailed response. I have tried with (ac02308). No netmap params and unpatched veth. iperf seems to work for a bit longer but it eventually halts and the connectivity is broken between veth1 and veth6.

I will try the options you mentioned while keeping an eye on the fixes that are being pushed.

**Comment by xuzhaokui** [comment]
> To overcome performance problems fo this setup, try to use the VALE switch in place of OVS.

@vmaffione I am also trying to combine ovs + veth + netmap，does [ovs-netmap patch](https://github.com/cnplab/ovs-netmap) work on this?

**Comment by vmaffione** [comment]
The ovs-netmap patch is not currently integrated. Unfortunately, I've never
tried it, personally, but is something we have to work on. You may ask the
original author (@micchie).

2016-02-12 7:49 GMT+01:00 xuzhaokui <notifications@github.com>:

> To overcome performance problems fo this setup, try to use the VALE switch
> in place of OVS.
>
> @vmaffione <https://github.com/vmaffione> I am also trying to combine ovs
> + veth + netmap，does ovs-netmap patch
> <https://github.com/cnplab/ovs-netmap> work on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luigirizzo/netmap/issues/118#issuecomment-183201846>.
>



-- 
Vincenzo Maffione

**Comment by micchie** [comment]
I will take a look into this tomorrow.

cheers,
- Michio

**Comment by xuzhaokui** [comment]
@micchie thanks a lot

**Comment by xuzhaokui** [comment]
@micchie I have successfully built ovs with the netmap-patch, after adding two internal ports to ovs, for example, `itnl1` and `itnl2`, I tried to run pkt-gen on `itnl1`:

> ./pkt-gen -f rx -i itnl1

and at the same time, I start sending packets to `itnl1` from `intl2`'s host stack(for example: ping from intl2). What I expected is that I can receive packets from `itnl1` by the tool pkt-gen.

however pkt-gen shows the error messge:

> 116.585532 nm_open [853] NIOCREGIF failed: Device or resource busy itnl1
> 116.585613 main [2046] Unable to open netmap:itnl1: Device or resource busy
> 116.585631 main [2114] aborting

PS: I have already inserted the netmap.ko and patched openvswitch.ko.
